storage/flash_area.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stddef.h>
+
+#include "flash.h"
+#include "flash_area.h"
+
+uint32_t flash_area_get_size(const flash_area_t *area) {
+  uint32_t size = 0;
+  for (int i = 0; i < area->num_subareas; i++) {
+    size += flash_sector_size(area->subarea[i].first_sector,
+                              area->subarea[i].num_sectors);
+  }
+  return size;
+}
+
+uint16_t flash_area_total_sectors(const flash_area_t *area) {
+  uint16_t total = 0;
+  for (int i = 0; i < area->num_subareas; i++) {
+    total += area->subarea[i].num_sectors;
+  }
+  return total;
+}
+
+int32_t flash_get_sector_num(const flash_area_t *area,
+                             uint32_t sector_inner_num) {
+  uint16_t sector = 0;
+  uint16_t remaining = sector_inner_num;
+  for (int i = 0; i < area->num_subareas; i++) {
+    if (remaining < area->subarea[i].num_sectors) {
+      sector = area->subarea[i].first_sector + remaining;
+      return sector;
+    } else {
+      remaining -= area->subarea[i].num_sectors;
+    }
+  }
+
+  return -1;
+}
+
+static secbool get_sector_and_offset(const flash_area_t *area, uint32_t offset,
+                                     uint16_t *sector_out,
+                                     uint32_t *offset_out) {
+  for (int i = 0; i < area->num_subareas; i++) {
+    // get subarea parameters
+    uint16_t first_sector = area->subarea[i].first_sector;
+    uint16_t num_sectors = area->subarea[i].num_sectors;
+    uint32_t size = flash_sector_size(first_sector, num_sectors);
+
+    // in this subarea?
+    if (offset < size) {
+      uint16_t found_sector = flash_sector_find(first_sector, offset);
+      *sector_out = found_sector;
+      *offset_out =
+          offset - flash_sector_size(first_sector, found_sector - first_sector);
+      return sectrue;
+    }
+
+    offset -= size;
+  }
+
+  return secfalse;
+}
+
+const void *flash_area_get_address(const flash_area_t *area, uint32_t offset,
+                                   uint32_t size) {
+  uint16_t sector;
+  uint32_t sector_offset;
+
+  if (!get_sector_and_offset(area, offset, &sector, &sector_offset)) {
+    return NULL;
+  }
+
+  return flash_get_address(sector, sector_offset, size);
```

Comment:
This function should return the address if the size+offset fits the whole subarea, not just one sector - (`flash_get_address` currently returns address if size+offset fits one sector on f4+unix, but all flash on u5, see also comment for that function). This behavior should be unified on all HWs. 

Before the refactoring, the u5's implementation of this function was behaving as it should, f4+unix wasn't - but since the sectors there are large it didn't cause problems.

One more thing to consider is security settings of the adjacent sectors on U5, should we succeed here if the sectors in subarea have different security? i think not, but it might be also reasonable to expect the subarea to have the same security in entire range.

Reply:
fixed, see 795fa67d4beff1d3de8cd9e43d4f214c6886d448